compiler/rustc_mir/src/monomorphize/polymorphize.rs

```diff
@@ -68,7 +76,13 @@ fn unused_generic_params(tcx: TyCtxt<'_>, def_id: DefId) -> FiniteBitSet<u32> {
     debug!("unused_generic_params: (after default) unused_parameters={:?}", unused_parameters);
 
     // Visit MIR and accumululate used generic parameters.
-    let body = tcx.optimized_mir(def_id);
+    let body = match context {
+        None => tcx.optimized_mir(def_id),
+        // FIXME(oli-obk): since this is solely used for codegen (I think?), should we keep using
+        // the optimized MIR for `const fn`? Need to adjust the above `is_mir_available` check
+        // in that case.
```

Comment:
I just wondered the same thing... this seems suspicious. "MIR for CTFE" AFAIK should never be monomorphized, so I don't think `mir_for_ctfe` should be called in this folder ever. But I know nothing about this code.^^

Reply:
It should be, a simple example would be
```rust
fn test<T: Trait> -> usize {
    T::ASSOC
}
```
here test can be polymorphized if `T::ASSOC` does not actually depend on `T`

Reply:
What is "polymorphization"? I thought all we are doing is "monomorphization"?

Reply:
polymorphization deduplicates instances of generic functions of their generic params don't matter and is currently disabled by default

Reply:
Oh... but why is this in the "monomorphization" folder?

Reply:
And if that is the case, doesn't that mean that for `const fn` we have to adjust *both* of their MIR? How can it be correct to only adjust one of them?

Or maybe it is just an internal part of monomorphization, but then again it should never be relevant for CTFE/Miri as those should not call monomorphization at all. It is not clear to me how things are glued together here, i.e., how does the output of polymorphization look like? Is it a new "level" on the MIR pipeline (mir_const, mir_optimized etc), or a new "item", or probably something else entirely?

Reply:
> so I don't think mir_for_ctfe should be called in this folder ever.

^ this is what "it should be" references

> Or maybe it is just an internal part of monomorphization,

That's mostly right, for more feel free to check out the relevant zulip wg https://rust-lang.zulipchat.com/#narrow/stream/216091-t-compiler.2Fwg-polymorphization. Polymorphization means that instead of emitting llvm-ir for `Vec<u32>::len` and `Vec<u64>::len` we may only emit llvm-ir for `Vec<_>::len` as that function doesn't actually depend on `T` (in the future... polymorphization is not clever enough for this yet and also currently disabled by default)

Reply:
> Polymorphization means that instead of emitting llvm-ir for Vec<u32>::len and Vec<u64>::len we may only emit llvm-ir for Vec<_>::len as that function doesn't actually depend on T (in the future... polymorphization is not clever enough for this yet and also currently disabled by default)

That sounds like it should never be relevant for Miri/CTFE, as they are unrelated to LLVM-IR generation. They work directly on polymorphic MIR.

So far everything you said seems to confirm that this code here should never use `mir_for_ctfe`.

Reply:
It is not relevant for `CTFE` but instead depends on the final value of constants. As we don't even generate `optimized_mir` for these constants we have to use `mir_for_ctfe` for them.

For `const fn` we should however use `optimized_mir`, as optimizations can remove mentions of generic args allowing for more deduplicatation. 

So while we should use `optimized_mir` for `const fn`, we don't have `optimized_mir` available for constants, for which we have to use `mir_for_ctfe` afaict.

Reply:
AFAIK codegen, including monomorphization, should never care about the MIR of constants. They should instead evaluate them.

Reply:
But polymorphization works on generic mir, so we can't just evaluate the constants

```rust
fn test<T, U>() -> usize {
    <Foo as Trait<T, U>>::ASSOC
}
```
Here `Foo::ASSOC` may depend on `T` but not on `U` so we can't just evaluate it while `test` is still generic, we do however want to still mark `U` as unused.

Reply:
Oh so it'll look at the MIR of the constant to figure out if it uses `U`?

Reply:
yes

Reply:
Ahhh I get it, thanks.

Then clearly it cannot just error on constants, but for functions it should always use the "runtime MIR" as monomorphization and thus also polymorphization are a runtime-code-only thing.